### PR TITLE
kernel: Explicitly define the main prototype if C++

### PR DESCRIPTION
**Pull request by aescolar**
If the embedded application has a `main()`, Zephyr requires it to have a signature with C linkage (no C++ name mangling).
Otherwise Zephyr's init will not call it, but will call the default weak stub.

But, when building with clang/llvm, when we build calling the compiler with `--frestanding` (for ex if CONFIG_MINIMAL_LIBC is set), the compiler will mangle the `main()` symbol name.
This is not incorrect behavior from the compiler, as, in principle, in a freestanding environment `main()` does not have a special meaning.
gcc is still not mangling it when called with `--frestanding`.
It is difficult to guess what other compilers will do; My guess is that all based on clang will mangle it.

To avoid this issue, we can define explicitly, in a header the application will always include, the `main` linkage as `extern C`.

    When building C++ code, let's explicitly define the prototype of the application main() function,
    to ensure the compiler does not mangle it, even if it was built freestanding.

    This fixes an issue with clang builds with
    freestanding, where a C++ application main()
    function name would be mangled otherwise.

### CODE REVIEW

**Review by stephanosio — DISMISSED** [review]
Just noting that only Clang mangles the name of `main` when compiling with `-ffreestanding`, which is not technically wrong since `main` is not anything special in a freestanding environment; but, it can be problematic as seen here.

**Review by galak — DISMISSED** [review]

**Review by andyross — COMMENTED** [review]
Not sure if kernel.h is the right spot for this.  That's a C header where we define Zephyr APIs.  This is a linker workaround applicable only to C++ builds.  My suggestion would be to put this somewhere under the toolchain.h tree, I guess, since it seems compiler-dependent?  Or somewhere specific to C++ if possible I guess.

Aesthetic complaint.  Would be nice to fix, doesn't rise to a -1.

(The explanation is also a little confusing: how exactly is main() getting mangled and why?  The special non-mangled linkage for main() is part of the language standard...)

**Comment by aescolar** [comment]
> (The explanation is also a little confusing: how exactly is main() getting mangled and why? The special non-mangled linkage for main() is part of the language standard...)

@andyross You mean the commit/PR description? I just updated the PR description.

> Not sure if kernel.h is the right spot for this. That's a C header where we define Zephyr APIs. This is a linker workaround applicable only to C++ builds. My suggestion would be to put this somewhere under the toolchain.h tree, I guess, since it seems compiler-dependent? Or somewhere specific to C++ if possible I guess.
> 
> Aesthetic complaint. Would be nice to fix, doesn't rise to a -1.

Thanks @andyross , I would be ok with putting this in the common toolchain header, if others would also agree:
In a way it does feel a bit out of place in kernel.h, but also in toolchain.h

Though at the same time, it does feel like this should be better present in kernel.h even without the `ifdef c++`. (But the drawback of doing that is that we'd force all apps to switch to that signature or get a warning/error).

**Comment by keith-packard** [comment]
> Though at the same time, it does feel like this should be better present in kernel.h even without the `ifdef c++`. (But the drawback of doing that is that we'd force all apps to switch to that signature or get a warning/error, so I agree with @stephanosio proposal of ifdef'ing it with c++). So I have the feeling at some point somebody will want it where it is now and to remove the ifdef.

Applications are now required to use this signature for main, so it's fine to place it in a global header that is intended to be included by the file defining the function. Both gcc and clang implicitly declare it when `-ffreestanding` is not in use, but having an explicit definition shouldn't cause any problems.

https://github.com/zephyrproject-rtos/zephyr/pull/54628/files#diff-38caa755f7300ce344017af71e0f454c0d43ee7cc2887e7c50e72c401c0b156f

File: include/zephyr/kernel.h [hunk 1]
```diff
@@ -35,6 +35,11 @@ BUILD_ASSERT(sizeof(int32_t) == sizeof(int));
 BUILD_ASSERT(sizeof(int64_t) == sizeof(long long));
 BUILD_ASSERT(sizeof(intptr_t) == sizeof(long));
 
+#ifdef __cplusplus
```
Comment (nashif) [inline include/zephyr/kernel.h:38]: I do not see how this relates to the kernel, can we find different place for this please?

**Review by nashif — DISMISSED** [review]

File: include/zephyr/kernel.h (hunk 1 above)
Comment (aescolar) [reply include/zephyr/kernel.h:38]: @nashif Would you be happy with toolchain.h as @andyross proposed?

**Comment by aescolar** [comment]
The best place I could think of is types.h
(toolchain* are macros headers, and did not feel right)
if somebody has a better proposal please say.

**Review by andyross — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Pull request merged**